core/src/main/scala/org/apache/spark/sql/SnappyContext.scala

```diff
@@ -544,41 +543,108 @@ class SnappyContext protected[spark](
+        s"Index and table have different databases " +
+          s"specified ${indexIdent.database} and ${tableIdent.database}")
+    }
     if (!catalog.tableExists(tableIdent)) {
       throw new AnalysisException(
-        s"$tableIdent is not an indexable table")
+        s"Could not find $tableIdent in catalog")
     }
-
-    //println("qualifiedTable=" + qualifiedTable)
+    import scala.collection.JavaConverters._
     catalog.lookupRelation(tableIdent) match {
-      case LogicalRelation(i: IndexableRelation, _) =>
-        i.createIndex(tableIdent.toString, sql)
+      case LogicalRelation(ir: JDBCMutableRelation, _) =>
+        ir.createIndex(indexIdent,
+          tableIdent,
+          indexColumns,
+          options)
+      case LogicalRelation(ir: JDBCAppendableRelation, _) =>
+        ir.createIndex(indexIdent,
+          tableIdent,
+          indexColumns,
+          options)
+        // Main table is updated to store the index information in it. We could have instead used
+        // createIndex method of HiveClient to do this. But, there were two main issues:
+        // a. Schema needs to be added to the HiveTable to supply indexedCols while creating
+        // the index. But, when schema is added to the HiveTable object, it throws a foreign
+        // key constraint failure on the serdes table. Need to look into this.
+        // b. The bigger issue is that the createIndex needs an indexTable for creating this
+        // index. Also, there are multiple things (like implementing HiveIndexHandler)
+        // that are hive specific and can create issues for us from maintenance perspective
+        catalog.alterTableToAddIndexProp(tableIdent, getIndexTable(indexIdent).toString)
+
       case _ => throw new AnalysisException(
         s"$tableIdent is not an indexable table")
     }
   }
 
+  private[sql] def getIndexTable(indexIdent: QualifiedTableName): QualifiedTableName = {
+    new QualifiedTableName(indexIdent.database,
+      ExternalStoreUtils.PREFIX_INDEX_TABLE + indexIdent.table)
+  }
+
+  private def constructDropSQL(indexName: String,
+                               ifExists : Boolean): String = {
+
+    val ifExistsClause = if (ifExists) "IF EXISTS" else ""
+
+    return s"DROP INDEX $ifExistsClause $indexName"
+  }
+
   /**
-   * Drop Index of a SnappyData table (created by a call to createIndexOnTable).
-   */
-  private[sql] def dropIndexOfTable(sql: String): Unit = {
+    * Drops an index on a table
+    * @param indexName Index name which goes in catalog
+    * @param ifExists Drop if exists, else exit gracefully
+    */
+  def dropIndexOnTable(indexName: String, ifExists: Boolean): Unit = {
+
+    val indexIdent = getIndexTable(catalog.newQualifiedTableName(indexName))
+
+    // Since the index does not exist in catalog, it may be a row table index.
+    if (!catalog.tableExists(indexIdent)) {
+      dropRowStoreIndex(indexName, ifExists)
+    } else {
+      catalog.lookupRelation(indexIdent) match {
```

Comment:
We need to track display of these column indexes in "show indexes", "select * from sys.indexes" in snappy-shell.